spectator/counter.py

```diff
@@ -1,50 +1,37 @@
-from abc import ABCMeta, abstractmethod
-from future.utils import with_metaclass
+from spectator.id import MeterId
+from spectator.spectatord import SpectatorD
 
-from spectator.atomicnumber import AtomicNumber
 
+class Counter:
+    """The value is the number of increments that have occurred since the last time it was
+    recorded. The value will be reported to the Atlas backend as a rate-per-second."""
 
-class AbstractCounter(with_metaclass(ABCMeta)):
+    meter_type = "c"
 
-    @abstractmethod
-    def increment(self, amount=1):
-        pass
+    def __init__(self, meter_id: MeterId, spectatord: SpectatorD = SpectatorD()) -> None:
+        self.meterId = meter_id
+        self._spectatord = spectatord
 
-    @abstractmethod
-    def count(self):
-        pass
-
-    @abstractmethod
-    def _measure(self):
-        pass
-
-
-class NoopCounter(AbstractCounter):
-
-    def increment(self, amount=1):
-        pass
+    def increment(self, amount: float = 1) -> None:
+        if amount > 0:
+            self._spectatord.publish(self.meter_type, self.meterId, amount)
 
-    def count(self):
+    @staticmethod
+    def count() -> float:
+        """Avoid breaking the API."""
         return 0
 
-    def _measure(self):
-        return {}
 
+class MonotonicCounter:
+    """The value is a monotonically increasing number. A minimum of two samples must be received
+    in order for spectatord to calculate a delta value and report it to the backend."""
 
-class Counter(AbstractCounter):
+    meter_type = "C"
 
-    def __init__(self, meterId):
-        self.meterId = meterId
-        self._count = AtomicNumber(0)
+    def __init__(self, meter_id: MeterId, spectatord: SpectatorD = SpectatorD()) -> None:
+        self.meterId = meter_id
+        self._spectatord = spectatord
 
-    def increment(self, amount=1):
+    def increment(self, amount: float = 0) -> None:
```

Comment:
`increment` doesn't really make sense for a monotonic counter. The only reason to use that type is if you are accessing a source that supplies a monotonically increasing value. As such the user typically isn't incrementing the value, but just specifying what it was when last sampled.

Also, depending on the source, the amount can be negative. For example, some of the system counters under `/proc` are signed integers and on overflow can be negative. The actual value doesn't matter, just the delta between samples over time.

Reply:
Changed this from `increment` to `set`.